Admittance and impedance vectors

stephane-caron opened the pull request:
This PR adds named types for admittance (motion → force) and impedance (force → motion) quantities. 

It allows one to write such things as:
```c++
sva::ForceVecd F = sva::ImpedanceVecd(10., 5.) * sva::MotionVecd({0., 0., 1.}, {0., 0., 0.});
sva::MotionVecd M = sva::AdmittanceVecd(0., 1.) * sva::ForceVecd({0., 0., 0.}, {5., 5., 0.});
```

Incidentally, it introduces a `moment()` alias to `couple()` in force vectors, along with a note to clarify that "couples" in SVA denote general moments and not [pure moments](https://en.wikipedia.org/wiki/Couple_(mechanics)).

Commit:
Added AdmittanceVec and ImpedanceVec types

Commit:
Homogeneous impedance/admittance constructors for convenience

Commit:
[minor] Conform to existing indentation style

Commit:
Added Zero() initializer to vector types

gergondet commented on src/SpaceVecAlg/AdmittanceVec.h:
```diff
@@ -0,0 +1,181 @@
+	typedef Vector3<T> vector3_t;
+	typedef Vector6<T> vector6_t;
+
+	friend class PTransform<T>;
```
Not needed

stephane-caron replied:
For now it is used below, same as in ``MotionVec<T>``. These typedefs could be removed from all types, but that would be a style point separate from this PR in my opinion.

gergondet replied:
I don't mean the `typedef` which are fine but rather the `friend` declaration

stephane-caron replied:
Ah, sorry, I misread your comment :) Done in eddef9a4a4983beb10cfa95a4f6ebb286c63d20f.

gergondet commented on src/SpaceVecAlg/AdmittanceVec.h:
```diff
@@ -0,0 +1,181 @@
+		linear_(vec.template tail<3>())
+	{}
+
+	/**
```
Missing a short description here and in the constructor above

stephane-caron replied:
Done in c88dad0398ad573c453821f2ab4cf7a791660d1f.

Note that these descriptions are missing in the original ``MotionVec<T>`` as well. (Not part of this PR though.)

gergondet commented on src/SpaceVecAlg/AdmittanceVec.h:
```diff
@@ -0,0 +1,181 @@
+	}
+
+public:
+	AdmittanceVec():
```
Missing doxygen

stephane-caron replied:
Done in d18080deb6bc9107ad2bc95bb925f61f9948bc4d

gergondet commented on src/SpaceVecAlg/AdmittanceVec.h:
```diff
@@ -0,0 +1,181 @@
+
+	bool operator!=(const AdmittanceVec<T>& av) const
+	{
+		return (angular_ != av.angular_) || (linear_ != av.linear_);
```
`return !(*this == av);`

stephane-caron replied:
Done in c1aea17de83c14922ec1499d46666ef017a44a76

gergondet reviewed (changes requested):
No significant problem imho.

Remarks I made about `AdmittanceVecd` implementation also apply to `ImpedanceVecd`

Two things to do:
- fix Windows CI (see https://github.com/jrl-umi3218/RBDyn/commit/c9fe688bcb5d77ec41b073a9efee70b4cb2f3db3)
- Implement Python bindings for `AdmittanceVecd` and `ImpedanceVecd`

Commit:
[ci] Update AppVeyor script

Commit:
Added empty constructors

See https://github.com/jrl-umi3218/SpaceVecAlg/pull/24#discussion_r212898182

Commit:
Complete constructor descriptions

See https://github.com/jrl-umi3218/SpaceVecAlg/pull/24/files/6c7287ee2d046a3e79ce54a7ae52d82e4ec21c22#r212898115

Commit:
Update != operator

See https://github.com/jrl-umi3218/SpaceVecAlg/pull/24/files/6c7287ee2d046a3e79ce54a7ae52d82e4ec21c22#r212900188

Commit:
Removed unused friend class

See https://github.com/jrl-umi3218/SpaceVecAlg/pull/24#discussion_r214630575

gergondet committed:
Add operator*= and operator/= for AdmittanceVec and ImpedanceVec

gergondet committed:
Add Python bindings related to #24

- AdmittanceVecd
- ImpedanceVecd
- New Zero() static method
- moment() alias for couple() in ForceVecd

gergondet commented:
I've added the Python bindings for the new types, new static functions and the new `moment()` alias.

Furthermore, I've also added `operator*=(T)` and `operator/=(T)` for `AdmittanceVec` and `ImpendanceVec`

Also, any reason why you didn't implement the symetric operation for `ImpedanceVec * MotionVec` and `AdmittanceVec * ForceVec` (i.e. `MotionVec * ImpedanceVec` and `ForceVec * AdmittanceVec` ?

stephane-caron commented:
Awesome, thanks Pierre!

> Also, any reason why you didn't implement the symetric operation for ImpedanceVec * MotionVec and AdmittanceVec * ForceVec (i.e. MotionVec * ImpedanceVec and ForceVec * AdmittanceVec ?

Hmm, no good reason, it's only because I'm used to writing them in that order :sweat_smile: We can definitely add the symmetric operations.

gergondet committed:
Symmetric operations Force*Admittance and Motion*Impedance

gergondet commented:
Added the symmetric operations then. Will merge when the build passes.

Thanks for your contribution!
